Fix #1556: Consolidate CollectionBackendApiService and WritableCollectionBackendApiService

joshuacano opened the pull request:
I have split and renamed CollectionBackendApiService and WritableCollectionBackendApiService to ReadOnlyCollectionBackendApiService and EditableCollectionBackendApiService. 

I need some assistance with helping me get all the requirements for this issue completed. So far I have done the following.

1. Split the CollectionBackendAPIService to EditableCollectionBackendApiService and ReadOnlyCollectionBackendApiService
2. Rename WritableCollectionDataHandler to EditableCollectionDataHandler
3. Add a get method to EditableCollectionDataHandler
4. Added tests in new file EditableCollectionBackendApiServiceSpec.js

Can I have some assistance in pointing out what else is left to be done for this issue? Thanks so much!

seanlip commented:
Thanks @joshuacano! As a PR, this looks great. Assigning to @BenHenning for advice on what to do next.

seanlip commented:
That said, the Travis checks seem to be failing with lint errors...

joshuacano commented:
two steps forward, one step back. I'll run the linter again locally see if it catches anything.

joshuacano commented:
Addressed lint issues, hopefully Travis checks pass.

BenHenning commented on core/controllers/collection_editor.py:
```diff
@@ -155,6 +155,44 @@ def _require_valid_version(self, version_from_payload, collection_version):
+        collection = (
+            collection_services.get_collection_by_id(collection_id))
+
+        if (collection is None or
```
Is this check needed when we use ``@require_editor``? I don't think so because that requires the user to have edit access to the collection, whereas we're only checking view access here. Users with edit access automatically have view access.

joshuacano replied:
You are correct, I will winnow the method to just update the values dict for now, and redirect when this method is called.

BenHenning replied:
Yeah, something seems a bit off with the new code. Please see my newer comment: https://github.com/oppia/oppia/pull/2434/files#r78504230

BenHenning commented on core/controllers/collection_editor.py:
```diff
@@ -155,6 +155,44 @@ def _require_valid_version(self, version_from_payload, collection_version):
+            self.redirect('/')
+            return
+
+        can_edit = (
```
Also, unless mistaken, this should always be true since ``@require_editor``

joshuacano replied:
Correcto.

BenHenning commented on core/templates/dev/head/collection_editor/CollectionEditor.js:
```diff
@@ -20,7 +20,7 @@
   'COLLECTION_DATA_URL_TEMPLATE', '/collection_handler/data/<collection_id>');
 oppia.constant(
-  'WRITABLE_COLLECTION_DATA_URL_TEMPLATE',
+  'COLLECTION_EDITABLE_DATA_URL_TEMPLATE',
```
Everywhere else you say "editable collection" rather than "collection editable". Prefer to be consistent.

BenHenning replied:
This was done.

BenHenning commented on core/templates/dev/head/domain/collection/EditableCollectionBackendApiService.js:
```diff
@@ -28,20 +28,22 @@
-        collectionDataUrl += '?allow_invalid_explorations=true';
+
+        // Does not work with following line
+        // collectionDataUrl += '?allow_invalid_explorations=true';
```
This should implicitly be true for the ``EditableCollectionDataHandler``. Ie, all collections provided by it should always support providing invalid explorations, and the read-only handler should never provide invalid explorations. This arg can be removed from the read-only handler and the editable handler should be updated appropriately.

BenHenning replied:
This was done.

BenHenning commented on core/templates/dev/head/domain/collection/EditableCollectionBackendApiService.js:
```diff
@@ -68,13 +70,14 @@ oppia.factory('WritableCollectionBackendApiService', [
           // The returned data is an updated collection dict.
           var collection = angular.copy(response.data);
 
           // Update the CollectionBackendApiService's cache with the new
```
Please update this comment; the service's name has changed.

BenHenning replied:
This was done.

BenHenning commented on core/templates/dev/head/domain/collection/ReadOnlyCollectionBackendApiService.js:
```diff
@@ -21,7 +21,7 @@
 // the collection instead. This file should not be included on the page in that
 // scenario.
-oppia.factory('CollectionBackendApiService', [
+oppia.factory('ReadOnlyCollectionBackendApiService', [
```
Can you please make sure the ``fileoverview`` specifically mentions this provides read-only collections?

BenHenning replied:
This was done.

BenHenning commented on core/templates/dev/head/domain/collection/ReadOnlyCollectionBackendApiServiceSpec.js:
```diff
@@ -13,11 +13,11 @@
-describe('Collection backend API service', function() {
-  var CollectionBackendApiService = null;
+describe('Read only collection backend API service', function() {
+  var ReadOnlyBackendApiService = null;
```
Prefer calling this ``ReadOnlyCollectionBackendApiService`` to maintain parity with the actual service name.

BenHenning replied:
This was done.

BenHenning commented on feconf.py:
```diff
@@ -416,7 +416,7 @@ def get_empty_ratings():
 ADMIN_URL = '/admin'
 COLLECTION_DATA_URL_PREFIX = '/collection_handler/data'
-COLLECTION_WRITABLE_DATA_URL_PREFIX = '/collection_editor_handler/data'
+COLLECTION_EDITABLE_DATA_URL_PREFIX = '/collection_editor_handler/data'
```
Ditto here (prefer to name things consistently)

BenHenning replied:
This was done.

BenHenning commented on core/templates/dev/head/domain/collection/EditableCollectionBackendApiServiceSpec.js:
```diff
@@ -0,0 +1,122 @@
+// Copyright 2015 The Oppia Authors. All Rights Reserved.
```
Nit: 2016

BenHenning replied:
This was done.

BenHenning commented on core/templates/dev/head/domain/collection/EditableCollectionBackendApiServiceSpec.js:
```diff
@@ -0,0 +1,122 @@
+// limitations under the License.
+
+/**
+ * @fileoverview Unit tests for EditableBackendApiService.
```
s/EditableBackendApiService/EditableCollectionBackendApiService

BenHenning replied:
This was done.

BenHenning commented on core/templates/dev/head/domain/collection/EditableCollectionBackendApiServiceSpec.js:
```diff
@@ -0,0 +1,122 @@
+ */
+
+describe('Editable collection backend API service', function() {
+  var EditableBackendApiService = null;
```
For parity, please name this ``EditableCollectionBackendApiService``

BenHenning replied:
This was done.

BenHenning commented on core/templates/dev/head/domain/collection/EditableCollectionBackendApiServiceSpec.js:
```diff
@@ -0,0 +1,122 @@
+    var successHandler = jasmine.createSpy('success');
+    var failHandler = jasmine.createSpy('fail');
+
+    $httpBackend.expect('GET', '/collection_handler/data/0').respond(
```
Shouldn't GET requests be going to ``collection_editor_handler``?

joshuacano replied:
Totally. They should. Thanks for catching this.

BenHenning commented on core/templates/dev/head/domain/collection/EditableCollectionBackendApiServiceSpec.js:
```diff
@@ -0,0 +1,122 @@
+
+    // Attempt Update to collection
+    EditableBackendApiService.updateCollection(
+      '0',
```
We generally don't use newlines to separate out multiple arguments, e.g.:

```javascript
EditableCollectionBackendApiService.updateCollection(
  '2', 'New Title', UndoRedoService.getChangeList()).then(successHandler, failHandler);
```

BenHenning replied:
This was changed.

BenHenning commented on core/templates/dev/head/domain/collection/EditableCollectionBackendApiServiceSpec.js:
```diff
@@ -0,0 +1,122 @@
+    $httpBackend.expect('PUT', '/collection_editor_handler/data/0').respond(
+      sampleDataResults);
+
+    // Attempt Update to collection
```
Perhaps: ``Send a request to update the collection``

BenHenning replied:
This was done.

BenHenning commented:
Thanks @joshuacano, I added comments. This definitely is looking good, but I think you need to add some backend tests since the backend data controller for the collection editor has changed. It now has support for retrieving collections as well as saving them. As I mentioned in one of my comments, there are also some changes needed for the data handler in the collection player, too (invalid collections should not be retrievable for the player).

joshuacano commented:
Sounds great thanks for all the useful feedback, I'll work on this and will
let you know once I'm done.

On Mon, Aug 29, 2016 at 9:32 PM, Ben Henning <notifications@github.com>
wrote:

> Thanks @joshuacano <https://github.com/joshuacano>, I added comments.
> This definitely is looking good, but I think you need to add some backend
> tests since the backend data controller for the collection editor has
> changed. It now has support for retrieving collections as well as saving
> them. As I mentioned in one of my comments, there are also some changes
> needed for the data handler in the collection player, too (invalid
> collections should not be retrievable for the player).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/oppia/oppia/pull/2434#issuecomment-243330358>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADKB7g4NAUFxuURgwxbqsTBxVyr6QiURks5qk7JcgaJpZM4Jtctp>
> .
>

joshuacano commented on core/controllers/collection_viewer.py:
```diff
@@ -80,10 +80,12 @@ def get(self, collection_id):
 class CollectionDataHandler(base.BaseHandler):
     """Provides the data for a single collection."""
 
+    @require_collection_playable
```
Added require_collection_playable to this handler, is this a requirement for readonly access to collection data?

BenHenning replied:
Good catch; you're right, only users with access should be able to play it.

BenHenning replied:
Since that decorator is used for the collection player itself, I think this is perfectly reasonable. Good catch.

joshuacano commented:
Have done the following.
1. made changes outlined
2. changed get method to return json instead of a rendered html template. 
3. added backend tests to test get and put functionality of controller.

I am getting a few failures on the e2e tests? But they do not seem to be related to the collection suite of services. Let me know if there are issues here.

joshuacano commented on core/controllers/collection_editor.py:
```diff
@@ -154,31 +140,30 @@ def _require_valid_version(self, version_from_payload, collection_version):
+            collection_dict = (
+                summary_services.get_learner_collection_dict_by_id(
+                    collection_id, self.user_id,
+                    allow_invalid_explorations=True))
```
I hardcoded allow_invalid_explorations=True here.

joshuacano commented on core/controllers/collection_viewer.py:
```diff
@@ -84,14 +84,11 @@ class CollectionDataHandler(base.BaseHandler):
                 summary_services.get_learner_collection_dict_by_id(
                     collection_id, self.user_id,
-                    allow_invalid_explorations=allow_invalid_explorations))
+                    allow_invalid_explorations=False))
```
I also hardcoded invalid_explorations to False here.

joshuacano commented on core/templates/dev/head/domain/collection/EditableCollectionBackendApiService.js:
```diff
@@ -30,21 +30,18 @@
           });
 
-        // Does not work with following line
-        // collectionDataUrl += '?allow_invalid_explorations=true';
```
Removed this switch as the read_only and editable controller handles allow_invalid_explorations implicitly.

seanlip commented:
Hi @joshuacano -- just wanted to give you a quick heads-up that @BenHenning is away for a bit, so apologies for the delay. He should be able to take another look at this when he gets back!

Also, I see only a lint failure on travis which should be easy to fix. Could you PTAL? The e2e tests are occasionally flaky but I believe a maintainer has restarted them, and they all currently pass.

Thanks!

joshuacano commented:
Sure thing. Have corrected that lint issue. Will await for @BenHenning's review

BenHenning commented on core/controllers/collection_editor.py:
```diff
@@ -95,30 +95,16 @@ class CollectionEditorHandler(base.BaseHandler):
-    # TODO(bhenning): Implement read-only version of the editor. Until that
-    # exists, ensure the user has proper permission to edit this collection
-    # before seeing the editor.
+    # Require editor handles all edit permissions for a collection
```
This is sort of obvious; is the comment needed?

joshuacano replied:
No, it is not, this has been removed.

BenHenning commented on core/controllers/collection_editor.py:
```diff
@@ -154,6 +140,43 @@ def _require_valid_version(self, version_from_payload, collection_version):
                 'which is too old. Please reload the page and try again.'
                 % (collection_version, version_from_payload))
 
+    # Require editor handles all edit permissions for a collection
```
Ditto for this comment.

joshuacano replied:
This has been removed.

BenHenning commented on core/controllers/collection_editor.py:
```diff
@@ -154,6 +140,43 @@ def _require_valid_version(self, version_from_payload, collection_version):
+        except Exception as e:
+            raise self.PageNotFoundException(e)
+
+        self.values.update({
```
All these values don't seem right to me; shouldn't this controller only be passing a collection dict to the frontend, just like it was before?

joshuacano replied:
I believe you are right, let me remove everything besides the collection dict.

joshuacano replied:
The Collection_dict as collection is now the only value returned.

BenHenning commented on core/controllers/collection_editor.py:
```diff
@@ -172,12 +195,19 @@ def put(self, collection_id):
+        self.values.update({
+            'collection_id': collection.id,
+            'collection': collection_dict,
+            'title': collection.title})
```
Nit: please put the ``})`` on its own line.

joshuacano replied:
This has been done.

BenHenning commented on core/controllers/collection_editor.py:
```diff
@@ -172,12 +195,19 @@ def put(self, collection_id):
-        # Send the updated collection back to the frontend.
-        self.values.update(collection_dict)
+        # return updated collection
+        self.values.update({
```
Why have ``collection_id`` and ``title`` here?

joshuacano replied:
I was referencing some earlier code, this has been removed and only collection is returned.

BenHenning commented on core/controllers/collection_editor.py:
```diff
@@ -172,12 +195,19 @@ def put(self, collection_id):
 
-        # Send the updated collection back to the frontend.
-        self.values.update(collection_dict)
+        # return updated collection
```
When adding comments, it's generally a good idea to keep them formatted in a consistent way with neighboring comments (e.g. capitalizing them and ending with punctuation, in this case).

joshuacano replied:
Got it, will note that for the future, this has been done.

BenHenning commented on core/controllers/collection_editor_test.py:
```diff
@@ -22,6 +22,7 @@
 import feconf
 
 
+
```
Please remove extra line.

joshuacano replied:
This has been done.

BenHenning commented on core/controllers/collection_editor_test.py:
```diff
@@ -99,6 +100,68 @@ def test_access_collection_editor_page(self):
         self.assert_can_edit(response.body)
         self.logout()
 
+
```
Should only have 1 line between tests in the same class, and 2 lines between classes.

joshuacano replied:
Righto, I overlooked this, This has been done

BenHenning commented on core/controllers/collection_editor_test.py:
```diff
@@ -99,6 +100,68 @@ def test_access_collection_editor_page(self):
+
+        self.logout()
+
+
```
Ditto (just 1 line)

joshuacano replied:
This has been done.

BenHenning commented on core/controllers/collection_editor_test.py:
```diff
@@ -99,6 +100,68 @@ def test_access_collection_editor_page(self):
+        self.assertEqual(
+            'A new title', json_response['title'])
+        self.logout()
+
```
Ditto (just one line)

joshuacano replied:
This has been done.

BenHenning commented on core/controllers/collection_editor_test.py:
```diff
@@ -99,6 +100,68 @@ def test_access_collection_editor_page(self):
+        self.assertEqual(self.COLLECTION_ID, json_response['collection_id'])
+        self.assertEqual(
+            'Introduction to Collections in Oppia', json_response['title'])
+
```
What about asserting the collection that's in ``json_response`` is correct?

joshuacano replied:
I rewrote this to assert that the title in the collection is correct.

BenHenning commented on core/controllers/collection_editor_test.py:
```diff
@@ -99,6 +100,68 @@ def test_access_collection_editor_page(self):
+                       self.COLLECTION_ID))
+        csrf_token = self.get_csrf_token_from_response(response)
+
+        json_dict = {"version" : 1,
```
Consider formatting a bit more consistently with the rest of the codebase:

```python
json_dict = {
    'version' : 1,
    'commit_message' : 'changed title',
    'change_list' : [{
        'cmd': 'edit_collection_property',
        'property_name': 'title',
        'new_value': 'A new title'
    }]
}
```

Also, we only use single quotes for strings.

joshuacano replied:
Got It, This has been corrected.

BenHenning commented on core/controllers/collection_editor_test.py:
```diff
@@ -99,6 +100,68 @@ def test_access_collection_editor_page(self):
+                       self.COLLECTION_ID), json_dict, csrf_token=csrf_token)
+
+        self.assertEqual(self.COLLECTION_ID, json_response['collection_id'])
+        self.assertEqual(2, json_response['collection']['version'])
```
Can you please assert ``['collection']['title']``? I don't think the title needs to be in the JSON response, since you can just assert it as part of the collection dict.

joshuacano replied:
Totally agree, this has been done.

BenHenning commented on core/controllers/collection_viewer.py:
```diff
@@ -80,23 +80,20 @@ def get(self, collection_id):
-            'can_edit': (
-                self.user_id and rights_manager.Actor(self.user_id).can_edit(
-                    feconf.ACTIVITY_TYPE_COLLECTION, collection_id)),
+            'can_edit': False,
```
Why did you change this line?

joshuacano replied:
I did not even consider that the can_edit attribute was used for a variety of things, will change back.

joshuacano replied:
This has been changed back

BenHenning commented on core/templates/dev/head/collection_editor/collection_editor.html:
```diff
@@ -93,7 +93,7 @@
   <script src="{{TEMPLATE_DIR_PREFIX}}/collection_editor/statistics_tab/CollectionStatisticsTabDirective.js"></script>
 
-  <script src="{{TEMPLATE_DIR_PREFIX}}/domain/collection/CollectionBackendApiService.js"></script>
+  <script src="{{TEMPLATE_DIR_PREFIX}}/domain/collection/ReadOnlyCollectionBackendApiService.js"></script>
```
Nit: please make sure this is in the right place (alphabetical sorting)

joshuacano replied:
This has been done.

BenHenning commented on core/templates/dev/head/collection_player/collection_player.html:
```diff
@@ -100,7 +100,7 @@ <h2 ng-if="collectionPlaythrough.hasFinishedCollection()" class="oppia-page-head
   <script src="{{TEMPLATE_DIR_PREFIX}}/collection_player/CollectionPlayer.js"></script>
   <script src="{{TEMPLATE_DIR_PREFIX}}/collection_player/CollectionNodeListDirective.js"></script>
-  <script src="{{TEMPLATE_DIR_PREFIX}}/domain/collection/CollectionBackendApiService.js"></script>
+  <script src="{{TEMPLATE_DIR_PREFIX}}/domain/collection/ReadOnlyCollectionBackendApiService.js"></script>
```
Ditto (alphabetize please)

joshuacano replied:
Done.

BenHenning commented on core/templates/dev/head/domain/collection/EditableCollectionBackendApiServiceSpec.js:
```diff
@@ -0,0 +1,120 @@
+
+    // Send a request to update collection
+    EditableCollectionBackendApiService.updateCollection(
+      '0', '2', 'New Title', UndoRedoService.getChangeList()).then(
```
Firstly, I don't think you need ``originalCol``. Also, consider cloning the collection returned and changing that, rather than changing ``sampleDataResults``. For instance:

```javascript
var collection = EditableCollectionBackendApiService.fetchCollection('0');
$httpBackend.flush();

collection.title = 'New Title';

$httpBackend.expect('PUT', '/collection_editor_handler/data/0').respond(
  collection);

// Send a request to update the collection
EditableCollectionBackendApiService.updateCollection(
  collection.id, collection.version, collection.title, []).then(
    successHandler, failHandler);
$httpBackend.flush();

...
```

joshuacano replied:
I'm not entirely sure if the new way I implemented the test  is the best practice for using promises (In my code you can see the way I implemented it using collectionFunc)

BenHenning commented on core/templates/dev/head/domain/collection/EditableCollectionBackendApiServiceSpec.js:
```diff
@@ -0,0 +1,120 @@
+        title: 'Collection Under Test',
+        category: 'Test',
+        objective: 'To pass',
+        schema_version: '1',
```
Did you mean ``version``?

joshuacano replied:
Correct, I have changed this.

BenHenning commented:
Thanks @joshuacano this is really coming together! I had a new batch of comments after your changes, but I think it's getting pretty close. Most of my comments were small style nits.

seanlip commented:
Hi @joshuacano, just a quick additional note -- when you address a comment in a successive commit, please could you add a reply saying "Done."? That will make it easier to keep track of what's been addressed and what hasn't. Thanks!

joshuacano commented:
@seanlip  Will definitely do that, makes it much clearer.

joshuacano commented:
@BenHenning I believe I have addressed all of your comments. I had one minor question about promises in the collection_editor_test, but other than that I believe the items are addressed. Thanks a ton for all the useful feedback again.

BenHenning commented on core/templates/dev/head/collection_player/collection_player.html:
```diff
@@ -99,10 +99,10 @@ <h2 ng-if="collectionPlaythrough.hasFinishedCollection()" class="oppia-page-head
   <script src="{{TEMPLATE_DIR_PREFIX}}/collection_player/CollectionPlayer.js"></script>
-  <script src="{{TEMPLATE_DIR_PREFIX}}/collection_player/CollectionNodeListDirective.js"></script>
-  <script src="{{TEMPLATE_DIR_PREFIX}}/domain/collection/CollectionBackendApiService.js"></script>
+  <script src="{{TEMPLATE_DIR_PREFIX}}/collection_player/CollectionNodeListDirective.js"></script>  
```
Please remove trailing spaces.

joshuacano replied:
This has been done.

BenHenning commented on core/controllers/collection_editor.py:
```diff
@@ -155,6 +140,25 @@ def _require_valid_version(self, version_from_payload, collection_version):
+            raise self.PageNotFoundException(e)
+
+        self.values.update({
+            'collection': collection_dict
```
Why not do: ``self.values.update(collection_dict)``?

Since there's only one element in the dict, we could just send over the dict in-place. See my comment below for the put method, as it's similar.

joshuacano replied:
I was following the same syntax as in collection_viewer.py from the CollectionDataHandler. Should we have them uniformly return the same attributes? Or should i leave the EditableCollectionDataHandler to return in a slightly different format? In Collection_viewer.py we have
`self.values.update({
            'can_edit': (
                self.user_id and rights_manager.Actor(self.user_id).can_edit(
                    feconf.ACTIVITY_TYPE_COLLECTION, collection_id)),
            'collection': collection_dict,
            'is_logged_in': bool(self.user_id),
            'session_id': utils.generate_new_session_id(),
        })`

BenHenning replied:
Discussed at TL meeting; you bring up great points. Let's keep it as-is because it is better to introduce a new pattern all-out rather than as part of a random PR. Thanks!

BenHenning commented on core/controllers/collection_editor.py:
```diff
@@ -172,12 +176,18 @@ def put(self, collection_id):
-        self.values.update(collection_dict)
+        # Return updated collection.
+        self.values.update({
+            'collection': collection_dict
```
I feel like the comment and update could be reverted back to what they were originally:

```python
# Send the updated collection back to the frontend.
self.values.update(collection_dict);
```

WDYT?

joshuacano replied:
I agree.

BenHenning replied:
Thanks for making the correct code change (despite my comment above). Can you make sure the comment is copied over, though, as suggested? Ie:

s/Return updated collection./Send the updated collection back to the frontend.

joshuacano replied:
This has been done.

BenHenning commented on core/controllers/collection_editor_test.py:
```diff
@@ -99,6 +99,69 @@ def test_access_collection_editor_page(self):
+
+        # Check that whitelisted users can access the data
+        # from the editable_collection_data_handler
+        self.login(self.EDITOR_EMAIL)
```
What about testing to ensure users who do not have access to the editor receive an error when calling put()?

BenHenning commented on core/controllers/collection_editor.py:
```diff
@@ -172,12 +176,18 @@ def put(self, collection_id):
 
         # Retrieve the updated collection.
+        collection = (
+            collection_services.get_collection_by_id(collection_id))
```
This variable doesn't seem to be used. Do we need it?

joshuacano replied:
We do not, removing.

BenHenning commented on core/templates/dev/head/domain/collection/EditableCollectionBackendApiServiceSpec.js:
```diff
@@ -0,0 +1,123 @@
+    var collectFunc = EditableCollectionBackendApiService.fetchCollection('0');
+    collectFunc.then(function(data) {
+      collection = data;
+    });
```
You could also do this without introducing a new variable:

```javascript
EditableCollectionBackendApiService.fetchCollection('0').then(
  function(data) {
    collection = data;
  });
```

joshuacano replied:
Great! I have changed code to use this.

seanlip commented:
Hi @joshuacano -- do you think you could resolve the remaining comments and merge develop into this branch, so that Ben can take another pass? Thanks!

(Also, @BenHenning, I think there's one open [comment](https://github.com/oppia/oppia/pull/2434#discussion_r79765343) that you might want to look at.)

joshuacano commented:
Yes for sure I will do that by tomorrow.

Joshua Cano

> On Sep 26, 2016, at 11:23, Sean Lip <notifications@github.com> wrote:
> 
> Hi @joshuacano -- do you think you could resolve the remaining comments and merge develop into this branch, so that Ben can take another pass? Thanks!
> 
> (Also, @BenHenning, I think there's one open comment that you might want to look at.)
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

BenHenning commented:
Hi @joshuacano, I addressed your unresolved question regarding the ``self.values.update()`` code. Please let me know if there's anything else you need to help resolve the remaining comments. Thanks!

joshuacano commented:
Great @BenHenning I believe I have resolved all questions. Still having a devil of a time with these e2e_tests, but I'm fairly sure it is my machine. Let me know if all the changes look good?

BenHenning commented on core/controllers/collection_editor_test.py:
```diff
@@ -99,6 +109,85 @@ def test_access_collection_editor_page(self):
+
+        # Check that Non_whitelisted users cannot access the put method
+        # from the editable_collection_data_handler.
+        self.login(self.OWNER_EMAIL)
```
This isn't quite what I meant. Can you please update this to test that users who do not have edit access to a given collection cannot use ``put()``? This access checking should not be checking that whitelisting works correctly: whitelisting will be removed in the future. This test should ensure people cannot update collections to which they do not have edit access (this is a really important test to have).

joshuacano replied:
I believe I have added the tests you were thinking of, ensuring editors have access and viewers (Non-editors) do not have access. I ended up modifying the can_access test as well. Let me know if this is more along the lines of what you were thinking.

joshuacano commented:
@BenHenning Completely correct about the collection_editor. That was an error in the merge, I should have caught that, apologies. I will correct the collection_editor_test.

BenHenning commented:
This looks excellent; thanks @joshuacano. LGTM!

BenHenning commented:
I restarted the Travis build; the last run seemed to have timed out.
